### PR TITLE
Text styles of account-list, account-names from webservice,beacon-node.service -> BehaviorSubject

### PR DESCRIPTION
**Pull request by mauser001**
**account-list;** 
- texts and icons in the table are now white (switched mat-light-theme with mat-dark-theme, and added color:white to body in the styles.scss)
- added the loading of the  account names via the webservice

**beacon-node.service:** 
- made a BeaconState object for holding all state relevant data as BehaviorSubject's. 
- beaconNodeEndpoint is now loaded only once 

**beacon-node-status**
- adapted to new BeaconState object

**gitignore**
- added Visual Studio project folder

### CODE REVIEW

File: src/app/modules/wallet/pages/account-list/account-list.component.ts [hunk 1]
```diff
@@ -53,14 +63,12 @@ export class AccountListComponent implements OnInit {
   }
 }
 
-function createNewUser(id: number): UserData {
-  const name = NAMES[Math.round(Math.random() * (NAMES.length - 1))] + ' ' +
-      NAMES[Math.round(Math.random() * (NAMES.length - 1))].charAt(0) + '.';
+function createNewUser(id: number, name: string): UserData {
```
Comment (rauljordan) [inline src/app/modules/wallet/pages/account-list/account-list.component.ts:66]: Can we remove all this mock logic and instead just use the accounts and display real data?

File: src/app/modules/wallet/pages/account-list/account-list.component.ts [hunk 2]
```diff
@@ -2,6 +2,9 @@ import { Component, OnInit, ViewChild } from '@angular/core';
 import { MatPaginator } from '@angular/material/paginator';
 import { MatSort } from '@angular/material/sort';
 import { MatTableDataSource } from '@angular/material/table';
+import { tap } from 'rxjs/operators';
+import { Account } from '../../../../proto/validator/accounts/v2/web_api';
```
Comment (rauljordan) [inline src/app/modules/wallet/pages/account-list/account-list.component.ts:6]:
```suggestion
import { Account } from 'src/app/proto/validator/accounts/v2/web_api';
```

File: src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.ts [hunk 3]
```diff
@@ -1,5 +1,8 @@
 import { Component } from '@angular/core';
-import { BeaconNodeService } from 'src/app/modules/core/services/beacon-node.service';
+import { BehaviorSubject } from 'rxjs';
+import { tap } from 'rxjs/operators';
+import { BeaconNodeService, BeaconState } from 'src/app/modules/core/services/beacon-node.service';
+import { ValidatorBalances } from '../../../../proto/eth/v1alpha1/beacon_chain';
```
Comment (rauljordan) [inline src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.ts:5]:
```suggestion
import { ValidatorBalances } from 'src/app/proto/eth/v1alpha1/beacon_chain';
```

File: src/app/modules/core/services/beacon-node.service.ts [hunk 4]
```diff
@@ -11,9 +11,20 @@ import {
...
 } from 'src/app/proto/validator/accounts/v2/web_api';
 
+import {
+  ValidatorBalances,
+} from 'src/app/proto/eth/v1alpha1/beacon_chain';
+
 const POLLING_INTERVAL = 3000;
 const BEACON_API_SUFFIX = '/eth/v1alpha1';
 
+export class BeaconState {
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:26]: Beacon state is confusing naming because it contradicts with the [`BeaconState`](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#beaconstate) data structure

Comment (rauljordan) [reply src/app/modules/core/services/beacon-node.service.ts:26]: Perhaps we should use something like BeaconNodeState ?

File: src/app/modules/core/services/beacon-node.service.ts [hunk 5]
```diff
@@ -11,9 +11,20 @@ import {
...
+  beaconNodeEndpoint$: BehaviorSubject<string> = new BehaviorSubject(undefined);
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:27]: Is behavior subject what we want here? Feels weird to initialize it with undefined or false. Perhaps `ReplaySubject` works best here

File: src/app/modules/core/services/beacon-node.service.ts [hunk 6]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
-
-  // Observables.
-  status$ = this.http.get<NodeConnectionResponse>(`${this.apiUrl}/health/node_connection`).pipe(
-    tap((res: NodeConnectionResponse) => {
-      this.beaconNodeEndpoint = res.beaconNodeEndpoint + BEACON_API_SUFFIX;
-    }),
-  );
+  beaconState: BeaconState = new BeaconState();
+
+  private get getBeaconEndpoint$(): BehaviorSubject<string> {
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:46]: We should avoid exposing behavior subjects, can we return it asObservable() instead?

File: src/app/modules/core/services/beacon-node.service.ts [hunk 7]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
+    if (!this.beaconState.beaconNodeEndpoint$.getValue()) {
+      this.http.get<NodeConnectionResponse>(`${this.apiUrl}/health/node_connection`).subscribe(
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:48]: Let's try to avoid any logic in the subscribe functions in the code and instead proceed with pure, functional-reactive methods in the services

File: src/app/modules/core/services/beacon-node.service.ts [hunk 8]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
+    }
+    return this.beaconState.beaconNodeEndpoint$;
+  }
+
+  private updateBalances(): void {
+    this.getBeaconEndpoint$.pipe(
+      filter((beacanNodeEnpoint) => {
+        return beacanNodeEnpoint !== undefined;
+      }),
+      map(beacanNodeEnpoint => {
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:64]: typo: `beacanNodeEnpooint`

File: src/app/modules/core/services/beacon-node.service.ts [hunk 9]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
+          }
+        );
+    }
+    return this.beaconState.beaconNodeEndpoint$;
+  }
+
+  private updateBalances(): void {
+    this.getBeaconEndpoint$.pipe(
+      filter((beacanNodeEnpoint) => {
+        return beacanNodeEnpoint !== undefined;
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:62]: No need to have this check if we use a replay subject

File: src/app/modules/core/services/beacon-node.service.ts [hunk 10]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
+      }),
+      map(beacanNodeEnpoint => {
+        return this.http.get<ValidatorBalances>(`${beacanNodeEnpoint}/validator/balances`)
+          .subscribe((result) => {
+            this.beaconState.balances$.next(result)
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:67]: Avoid logic in the subscribe func. Why can't balances just be an observable instead? We can use switch map

File: src/app/modules/core/services/beacon-node.service.ts [hunk 11]
```diff
@@ -25,16 +36,41 @@ export class BeaconNodeService {
...
+          });
+      })
+    ).subscribe();
+  }
+
   statusPoll$ = interval(POLLING_INTERVAL).pipe(
     startWith(0),
-    mergeMap(_ => this.status$),
-  );
+    tap(() => this.updateBalances()),
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:96]: Let's avoid side-effects like these

File: src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.html [hunk 12]
```diff
@@ -1,31 +1,31 @@
-<mat-card class="bg-paper" *ngIf="(nodeConnection$ | async) as conn">
+<mat-card class="bg-paper">
   <div class="flex items-center">
     <div>
       <div
         class="pulsating-circle"
-        [class.green]="conn.connected"
-        [class.red]="!conn.connected"></div>
+        [class.green]="(beaconState.connected$ | async)"
```
Comment (rauljordan) [inline src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.html:6]: Seems like the logic became more verbose than just doing a single async pipe at the beginning

**Review by rauljordan — CHANGES_REQUESTED** [review]

File: src/app/modules/core/services/beacon-node.service.ts (hunk 4 above)
Comment (mauser001) [reply src/app/modules/core/services/beacon-node.service.ts:26]: sure i will rename it

File: src/app/modules/core/services/beacon-node.service.ts (hunk 5 above)
Comment (mauser001) [reply src/app/modules/core/services/beacon-node.service.ts:27]: the advantage of a BehaviorSubject is that it keeps beeing accessable after it changed ... think this is needed if we only want to load the endpoint once

File: src/app/modules/core/services/beacon-node.service.ts [hunk 13]
```diff
@@ -25,16 +41,58 @@ export class BeaconNodeService {
...
+  }
+
+  private updateBalances(): void {
+    this.getBeaconEndpoint$.pipe(
+      filter((beacanNodeEnpoint) => {
+        return beacanNodeEnpoint !== undefined;
+      }),
+      map(beacanNodeEnpoint => {
+        return this.http.get<ValidatorBalances>(`${beacanNodeEnpoint}/validators/balances`)
+          .subscribe((result) => {
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:66]: This sub isn't cleaned up

File: src/app/modules/wallet/pages/account-list/account-list.component.ts [hunk 14]
```diff
@@ -30,12 +33,19 @@ export class AccountListComponent implements OnInit {
...
-  constructor() {
+  constructor(private walletService: WalletService) {
     // Create 100 users.
-    const users = Array.from({length: 100}, (_, k) => createNewUser(k + 1));
+    //const users = Array.from({length: 100}, (_, k) => createNewUser(k + 1));
 
     // Assign the data to the data source for the table to render.
-    this.dataSource = new MatTableDataSource(users);
+
+    walletService.accounts$.pipe(
```
Comment (rauljordan) [inline src/app/modules/wallet/pages/account-list/account-list.component.ts:42]: This subscription is not cleaned up and will memory leak. Let's use `takeUntil(this.destroyed$)` as a pattern we use in other similar components: https://github.com/prysmaticlabs/prysm-web-ui/blob/c0ad24bb456fc43ddd5ca563878780c6c3330bda/src/app/modules/auth/login/login.component.ts#L63

File: src/app/modules/core/services/beacon-node.service.ts [hunk 15]
```diff
@@ -25,16 +41,58 @@ export class BeaconNodeService {
...
+            this.beaconNodeState.connected$.next(true);
+          },
+          (err) => {
+            this.beaconNodeState.connected$.next(false);
+          });
+      })
+    ).subscribe();
+  }
+
+  private updateSyncing(): void {
```
Comment (rauljordan) [inline src/app/modules/core/services/beacon-node.service.ts:77]: Where is this called? Wouldn't an observable polling pattern work best here?

```js
status$ = this.http.get<NodeConnectionResponse>(`${this.apiUrl}/health/node_connection`);
statusPoll$ = interval(POLLING_INTERVAL).pipe(
    startWith(0),
    mergeMap(_ => this.status$),
);
```
Where `statusPoll$` is used by the beacon-node-status component to continuously display changes in node status, such as when the beacon node disconnects, or when it finishes syncing, etc.

File: src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.html (hunk 12 above)
Comment (mauser001) [reply src/app/modules/dashboard/components/beacon-node-status/beacon-node-status.component.html]: The Idea was that the beaconNodeState is just a container for all the different informations ... if you just need the connected state somewhere, you could get it with this.nodeService.beaconNodeState.connected$ 
Otherwise you would always have to update the whole container object ... with this sollution you could have different intervals and only update specific infortmaions.

File: src/app/modules/core/services/beacon-node.service.ts (hunk 11 above)
Comment (mauser001) [reply src/app/modules/core/services/beacon-node.service.ts:96]: you mean the , at the end of "tap(() => this.updateBalances()),"?

File: src/app/modules/core/services/beacon-node.service.ts (hunk 8 above)
Comment (mauser001) [reply src/app/modules/core/services/beacon-node.service.ts:82]: corrected

File: src/app/modules/wallet/pages/account-list/account-list.component.ts [hunk 16]
```diff
@@ -2,40 +2,37 @@ import { Component, OnInit, ViewChild } from '@angular/core';
...
+  dataSource: MatTableDataSource<Account>;
 
   @ViewChild(MatPaginator, {static: true}) paginator: MatPaginator;
   @ViewChild(MatSort, {static: true}) sort: MatSort;
 
-  constructor() {
+  constructor(private walletService: WalletService) {
     // Create 100 users.
-    const users = Array.from({length: 100}, (_, k) => createNewUser(k + 1));
+    //const users = Array.from({length: 100}, (_, k) => createNewUser(k + 1));
```
Comment (rauljordan) [inline src/app/modules/wallet/pages/account-list/account-list.component.ts:23]:
```suggestion
```

File: src/app/modules/wallet/pages/account-list/account-list.component.spec.ts [hunk 17]
```diff
@@ -2,19 +2,35 @@ import { async, ComponentFixture, TestBed } from '@angular/core/testing';
...
   let component: AccountListComponent;
   let fixture: ComponentFixture<AccountListComponent>;
-
+  let service: WalletService = MockService(WalletService);
+  service.accounts$ = of({
+    accounts: [{
+      validatingPublicKey: null,
+      accountName: 'Fritz',
+      depositTxData: null,
+      derivationPath: "somepath"
```
Comment (rauljordan) [inline src/app/modules/wallet/pages/account-list/account-list.component.spec.ts:19]:
```suggestion
      derivationPath: 'somepath'
```

File: src/app/modules/core/pipes/uint-array-to-hex-string.pipe.ts [hunk 18]
```diff
@@ -0,0 +1,14 @@
+import { Pipe, PipeTransform } from '@angular/core';
+
+@Pipe({
+  name: 'uintArrayToHexString'
+})
+export class UintArrayToHexStringPipe implements PipeTransform {
```
Comment (rauljordan) [inline src/app/modules/core/pipes/uint-array-to-hex-string.pipe.ts:6]: This should have comprehensive tests, but we can do it in a separate PR as this one has been open for a bit

**Review by rauljordan — APPROVED** [review]
